fix: de- and encode bodies with utf-8

thomasbach-dev opened the pull request:
The previous implementation using `B8.unpack` and `B8.pack` handled UTF-8 characters wrong. For
example, using the character `î` (i-circumflex) in the generated stripeapi library I got:

Invalid UTF-8 characters found in POST body

as `B8.pack` would not convert the character into UTF-8.

I changed the implementation to use `encodeUtf8` and `decodeUtf8` from `Data.Text.Encoding`.

This is the upstream fix for https://github.com/Haskell-OpenAPI-Code-Generator/Stripe-Haskell-Library/issues/9

thomasbach-dev commented:
@joelfisch Should I also update the generated code in the example directory? If so, how should I do this?

joelfisch commented:
Thank you for creating the PR.

It would be helpful if you could update the `example` folder with the new code. You can just run the generator with the options `-o example/generatedCode specifications/petstore.yaml` but be aware that the repository contains a default `openapi-configuration.yml` which will be picked up automatically if you do not rename or remove it (or provide another configuration). Maybe it would be helpful if we created a file called `example-configuration.yml` or something like this which contains the following content:

```yaml
specification: specifications/petstore.yaml
outputDir: example/generatedCode
```

Then you could just run the generator with `--configuration example-configuration.yml`. Please note that the generated code will contain many other changes as well because I neglected the updating of the example in the past.

Another thing: The golden files are currently not equivalent to the code generated (or in the case of `Common.hs` just copied). You can update this using `.circleci/test_golden update` to get the exact result.

Thank you for your help!

thomasbach-dev commented:
I think I addressed all things mentioned above. I reversed the commit history to address your issues first and put my changes on top of them. I also added a small section to the README with instructions how to update examples and golden tests.

joelfisch commented on README.md:
````diff
@@ -70,3 +70,20 @@ The following features are not supported
+As these assume to be run as root you have to manually execute the commands listed there, e.g.:
+
+``` bash
+cd .circleci/golden
+openapi3-code-generator-exe ../specifications/z_complex_self_made_example.yml --output-dir expected
````
You can actually use the following snippet which I think is a little bit easier:

```bash
stack build # just to make sure the built executable is up-to-date
.circleci/test_golden update
```

thomasbach-dev replied:
Ah yeah. I saw that `[[ -d /root/... ` line in the script and something short-circuited in my brain telling me that this would not be apt to execute outside of CI without really trying it. :D
I'll adapt the section

joelfisch commented on example-configuration.yml:
```diff
@@ -0,0 +1,29 @@
+specification: specifications/petstore.yaml
+outputDir: example/generatedCode
+packageName: openapi
```
Just FYI: You do (or at least should) not have to list every property in the configuration file (only the ones which deviate from the default).

thomasbach-dev replied:
Yeah, I already thought so. :)
I just copied the `openapi-configuration.yml` file and made the necessary changes that seemed necessary. Not sure if these make sense. I can adapt the config to only contain the minimum necessary stuff.

joelfisch reviewed (approved):
Thank you so much for your help!

thomasbach-dev commented:
Ok, I incorporated your suggestions above.
Note that I also added `*.hs-boot` files in the `example` directory. Ping me if these should be dropped again and rather be ignored.

joelfisch commented:
> Ok, I incorporated your suggestions above. Note that I also added `*.hs-boot` files in the `example` directory. Ping me if these should be dropped again and rather be ignored.

That is absolutely correct. This looks good and I'll merge it. Thanks!
